### DISCUSSION TITLE
websocket support in application sdk

### DISCUSSION THREAD

**Discussion by vlotech** [edgexfoundry/community#172] [Q&A]
For realtime feedback of one of our applications, we would like to use a websocket and some web pages hosted from the application. 

As an example, we've used the camera-management example. This fits most of our purposes, but using a websocket seems to be impossible. 
The app-functions-sdk-go uses the http.TimeoutHandler which doesn't support Hijacking. The upgrade to websocket fails because of this. 

We've also looked at using the edgex-ui-server as example, but we could figure out how to add the message bus to it in the right way. 

What would be the best way forward?

**Comment by lenny-goodell** [comment]
@vlotech , yes the SDK uses the http.TimeoutHandler in the configuration for the internal WebServer. Why does this cause issue with adding `websocket` capability? 
You should add the `websocket` capability separate from the SDKs internal webserver. i.e. on a different port.

**Reply by vlotech** [reply]
That's how we solved it for now. 
In this application we'll also host an ui (which uses the websocket), but also some REST api calls. 
We could add the REST api to the extra webserver instance, or use two ports / urls and have to manage CORS. 

If that's the preferred design pattern, why forcefully have this internal webserver instead of making it optional? 

I'm really curious about the reasoning behind the design choice, and if the addition of a websocket was never concidered.

**Reply by lenny-goodell** [reply]
You didn't answer my question on why use of `http.TimeoutHandler` causes issue with with adding websocket capability?

**Reply by vlotech** [reply]
http.TimeoutHandler does not support the Hijacker interface which is needed to upgrade to an websocket

> TimeoutHandler supports the Pusher interface but does not support the Hijacker or Flusher interfaces.

**Reply by lenny-goodell** [reply]
Got it. 

The App SDK is deigned to develop microservices for the edge which process the sensor data, not for hosting a UI. It provides the ability  to expand on the small REST API (/version, /config, /ping, /secret) that a separate UI can utilize. The EdgeX UI utilizes the REST API of the EdgeX Services.

**Reply by vlotech** [reply]
Ok, That clear. We've used the Application SDK to build a UI. Simply because we needed the real-time data and it looked like the shortest route to success. 
The edge-UI example didn't have a message bus connection and this was the clearest route forward for us. 
We're new to Go and EdgeX and couldn't find a better way. 
Thanks for the feedback and if a cleaner approach would be possible with an example.. I'm open for it!!

**Reply by lenny-goodell** [reply]
I think you can use the Application SDK for this with the knowledge it wasn't designed for how you are using it. Thus you will have some work arounds like websocket capability hosted on a different port.
